src/devices/cpu/z180/z180asci.cpp

```diff
@@ -0,0 +1,287 @@
+}
+
+//-------------------------------------------------
+//  device_start - device-specific startup
+//-------------------------------------------------
+
+void z180asci_channel::device_start()
+{
+	save_item(NAME(m_asci_cntla));
+	save_item(NAME(m_asci_cntlb));
+	save_item(NAME(m_asci_stat));
+	save_item(NAME(m_asci_tdr));
+	save_item(NAME(m_asci_rdr));
+
+	transmit_register_reset();
+	receive_register_reset();
+
+	cpu_device *cpu = reinterpret_cast<cpu_device *>(owner());
+	if (m_id==0) {
+		cpu->state_add(Z180_CNTLA0,     "CNTLA0",    m_asci_cntla);
+		cpu->state_add(Z180_CNTLB0,     "CNTLB0",    m_asci_cntlb);
+		cpu->state_add(Z180_STAT0,      "STAT0",     m_asci_stat);
+		cpu->state_add(Z180_TDR0,       "TDR0",      m_asci_tdr);
+		cpu->state_add(Z180_RDR0,       "RDR0",      m_asci_rdr);
+	} else {
+		cpu->state_add(Z180_CNTLA1,     "CNTLA1",    m_asci_cntla);
+		cpu->state_add(Z180_CNTLB1,     "CNTLB1",    m_asci_cntlb);
+		cpu->state_add(Z180_STAT1,      "STAT1",     m_asci_stat);
+		cpu->state_add(Z180_TDR1,       "TDR1",      m_asci_tdr);
+		cpu->state_add(Z180_RDR1,       "RDR1",      m_asci_rdr);
+	}
+}
+
+//-------------------------------------------------
+//  device_reset - device-specific reset
+//-------------------------------------------------
+
+void z180asci_channel::device_reset()
+{
+	m_asci_cntla = 0;
+	m_asci_cntlb = 7; // SS2,SS1 and SS0 are 1 after reset
+	m_asci_tdr = 0;
+	m_asci_rdr = 0;
+	if (m_id==0) {
+		m_asci_cntla = (m_asci_cntla & Z180_CNTLA0_MPBR_EFR) | Z180_CNTLA0_RTS0;
+		m_asci_cntlb = (m_asci_cntlb & (Z180_CNTLB0_MPBT | Z180_CNTLB0_CTS_PS)) | 0x07;
+		m_asci_stat = m_asci_stat & (Z180_STAT0_DCD0 | Z180_STAT0_TDRE);
+	} else {
+		m_asci_cntla = (m_asci_cntla & Z180_CNTLA1_MPBR_EFR) | Z180_CNTLA1_CKA1D;
+		m_asci_cntlb = (m_asci_cntlb & Z180_CNTLB1_MPBT) | 0x07;
+		m_asci_stat = Z180_STAT1_TDRE;
+	}
+}
+
+uint8_t z180asci_channel::cntla_r()
+{
+	uint8_t data = m_asci_cntla;
+	LOG("Z180 CNTLA%d rd $%02x\n", m_id, data);
+	return data;
+}
+
+uint8_t z180asci_channel::cntlb_r()
+{
+	uint8_t data = m_asci_cntlb;
+	LOG("Z180 CNTLB%d rd $%02x\n", m_id, data);
+	return data;
+}
+
+uint8_t z180asci_channel::stat_r()
+{
+	uint8_t data = m_asci_stat;
+	LOG("Z180 STAT%d  rd $%02x\n", m_id, data);
+	return data;
+}
+
+uint8_t z180asci_channel::tdr_r()
+{
+	uint8_t data = m_asci_tdr;
+	LOG("Z180 TDR%d   rd $%02x\n", m_id, data);
+	return data;
+}
+
+uint8_t z180asci_channel::rdr_r()
+{
+	uint8_t data = m_asci_rdr;
+	LOG("Z180 RDR%d   rd $%02x\n", m_id, data);
+	m_asci_stat &= ~0x80;
+
+	return data;
+}
+
+void z180asci_channel::cntla_w(uint8_t data)
+{
+	LOG("Z180 CNTLA%d wr $%02x\n", m_id, data);
+	m_asci_cntla = data;
+	set_data_frame(1, (data & 4) ? 8 : 7, (data & 2) ? PARITY_ODD : PARITY_NONE, (data & 2) ? STOP_BITS_2 : STOP_BITS_1);
+}
+
+void z180asci_channel::cntlb_w(uint8_t data)
+{
+	LOG("Z180 CNTLB%d wr $%02x\n", m_id, data);
+	m_asci_cntlb = data;
+
+	int divisor = 1<<(m_asci_cntlb & 0x07);
```

Comment:
Here’s another case where one binary operator randomly doesn’t have spaces around it, but the others do.